### PR TITLE
Supervised Learning session

### PR DESCRIPTION
**Pull request by Paulj1989**

### CODE REVIEW

File: intermediate-skills-sessions/20-supervised-learning/index.qmd [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+rf_model.fit(X_train, y_train)
+
+# evaluate on test set
+rf_pred = rf_model.predict(X_test)
+rf_accuracy = accuracy_score(y_test, rf_pred)
+
+print(f"Random Forest Accuracy: {rf_accuracy:.1%}")
+```
+
+Random forests typically outperform single decision trees because the ensemble smooths out individual tree errors. The trade-off is reduced interpretability. You can't easily visualise 100 trees or explain a prediction as a single decision path.
```
Comment (jonaswillemsen) [inline intermediate-skills-sessions/20-supervised-learning/index.qmd:204]: This one hasn't!

File: intermediate-skills-sessions/20-supervised-learning/index.qmd [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+### Gradient Boosting
+
+Gradient boosting takes a different approach to combining trees. Instead of training trees independently and averaging, it trains them sequentially. This is an ensembling method called boosting. Each new tree focuses on the mistakes made by previous trees, gradually improving the model's performance.
+
+The process works as follows: 
+
+1. Train the first tree on the original data
+2. Identify where it makes prediction errors
+3. Ttrain the second tree to correct those errors
```
Comment (jonaswillemsen) [inline intermediate-skills-sessions/20-supervised-learning/index.qmd]: typo

File: intermediate-skills-sessions/20-supervised-learning/index.qmd [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+# create faceted plot
+g = sns.FacetGrid(importance_data, col='Model', height=4, aspect=0.8)
+g.map_dataframe(sns.barplot, y='Feature', x='Importance', order=feature_names)
+g.set_axis_labels('Importance', '')
+g.set_titles(col_template='{col_name}')
+plt.tight_layout()
+plt.show()
+```
+
```
Comment (jonaswillemsen) [inline intermediate-skills-sessions/20-supervised-learning/index.qmd:308]: The models have performed surprisingly closely for this dataset - do we need to explain any of that?

File: intermediate-skills-sessions/20-supervised-learning/index.qmd [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+rf_accuracy = accuracy_score(y_test, rf_pred)
+
+print(f"Random Forest Accuracy: {rf_accuracy:.1%}")
+```
+
+Random forests typically outperform single decision trees because the ensemble smooths out individual tree errors. The trade-off is reduced interpretability. You can't easily visualise 100 trees or explain a prediction as a single decision path.
+
+### Gradient Boosting
+
+Gradient boosting takes a different approach to combining trees. Instead of training trees independently and averaging, it trains them sequentially. This is an ensembling method called boosting. Each new tree focuses on the mistakes made by previous trees, gradually improving the model's performance.
```
Comment (jonaswillemsen) [inline intermediate-skills-sessions/20-supervised-learning/index.qmd]: Do we just need a footnote defining what "ensembling" means? It's somewhat implied from context but might help to clarify as it's a key concept

**Review by jonaswillemsen — COMMENTED** [review]
Overall I think this is a brilliant session - I'm not great on machine learning myself and I found it easy to follow. Just some minor comments.

File: intermediate-skills-sessions/20-supervised-learning/index.qmd (hunk 4 above)
Comment (Paulj1989) [reply intermediate-skills-sessions/20-supervised-learning/index.qmd]: Yeah I think this is where the structure of the session feels most muddled. I want to give enough information about each algorithm that everyone will broadly understand how they build on each other and get why some outperform others, but I don't think there will be the time to go into a ton of detail.

Either way, I think some explanation of ensembling methods is necessary to help lay the groundwork for random forests and gradient boosting models. I'm less sure about whether I should explain bagging/boosting or just skip over it entirely.

File: intermediate-skills-sessions/20-supervised-learning/index.qmd (hunk 3 above)
Comment (Paulj1989) [reply intermediate-skills-sessions/20-supervised-learning/index.qmd:308]: I will add a footnote to clarify that the dataset isn't that complex so the differences are relatively minimal.

File: intermediate-skills-sessions/20-supervised-learning/index.qmd (hunk 1 above)
Comment (Paulj1989) [reply intermediate-skills-sessions/20-supervised-learning/index.qmd:204]: Haha. Completely missed this, but good point. I have added in a couple more features to allow for a little extra complexity.

**Review by jonaswillemsen — APPROVED** [review]
Looks good.
